doc/contrib/new_api.md

```diff
@@ -0,0 +1,510 @@
+{
+    ElektraKey * key = elektraKeysetGet (ks, index);
+    elektraKeysetRemove (ks, index);
+    return key;
+}
+
+// Migration Note: replacement for ksLookup(ks, key, KDB_O_POP) would be
+/*
+elektraKeysetCut (ks, elektraKeysetLookup (ks, elektraKeyGetName (key)));
+*/
+
+// cuts arbitrary range out of keyset
+ElektraKeyset * elektraKeysetCutRange (ElektraKeyset * ks, size_t start, size_t end)
+{
+    ElektraKeyset * range = elektraKeysetGetRange (ks, start, end);
+    elektraKeysetRemoveRange (ks, start, end);
+    return range;
+}
+
+// cuts hierarchy out of keyset
+// Migration Note: replaces current ksCut
+ElektraKeyset * elektraKeysetCutHierarchy (ElektraKeyset * ks, const ElektraKeyname * root)
+{
+    size_t end;
+    size_t start = elektraKeysetFindHierarchy (ks, root, &end);
+    return elektraKeysetCutRange (ks, start, end);
+}
+```
+
+### Builders
+
+```c
+#define ELEKTRA_KEYSET(...) (elektraKeysetBuild ((ElektraKey *[]){__VA_ARGS__}, sizeof((ElektraKey *[]){__VA_ARGS__})/sizeof(ElektraKey*)))
+
+#define ELEKTRA_KEYNAME(ns_, name_) (ElektraKeyname){.ns = (ns_), .name = (name_), .size = sizeof (name_)}
+#define ELEKTRA_METANAME(name) ELEKTRA_KEYNAME(ELEKTRA_NS_META, (name))
+#define ELEKTRA_KEYVALUE_STRING(s) (ElektraKeyvalue){.value = (s), .size = strlen((s))}
+#define ELEKTRA_KEYVALUE_PTR(v) (ElektraKeyvalue){.value = &(v), .size = sizeof((v))}
+
+/**
+ * CAUTION releases reference to @p meta
+ */
+ElektraKey * elektraKeyBuild (const ElektraKeyname * name, const ElektraKeyvalue * value, ElektraKeyset * meta)
+{
+    ElektraKey * key = elektraKeyNew (name);
+    elektraKeySetValue (key, value);
+    elektraKeysetInsertAll ((ElektraKeyset *)elektraKeyGetMeta (key), meta);
+    elektraKeysetRelease (meta);
+    return key;
+}
+
+ElektraKey * elektraKeyBuildMeta (const ElektraKeyname * name, const ElektraKeyvalue * value)
+{
+    if (name->ns != ELEKTRA_NS_META)
+    {
+        return NULL;
+    }
+    return elektraKeyBuild (name, value, NULL);
+}
+
+/**
+ * DOES NOT retain new references to @p keys, instead moves existing reference into returned keyset
+ */
+ElektraKeyset * elektraKeysetBuild (ElektraKey ** keys, size_t size)
+{
+    ElektraKeyset * ks = elektraKeysetNew (size);
+
+    // build fake ElektraKeyset with everything that elektraKeysetInsertAll needs
+    // depending on elektraKeysetInsertAll this may require copying/sorting keys
+    ElektraKeyset fake = {
+        .data = &(struct ElektraKeysetDataCow) {
+            .array = keys,
+            .size = size,
+        }
+    };
+    elektraKeysetInsertAll (ks, &fake);
+
+    return ks;
+}
+```
+
+#### Usage Examples
+
+```c
+int a = 7;
+
+ElektraKey * key = elektraKeyNew (&ELEKTRA_KEYNAME(ELEKTRA_NS_SYSTEM, "foo")); // refs = 1
+
+ElektraKeyset * ks = ELEKTRA_KEYSET (
+    // elektraKeysetBuild consumes reference, doesn't retain itself
+    // we call elektraKeyRetain, so we can keep our own reference
+    elektraKeyRetain (key), // refs = 2
```

Comment:
Doesn't look easy to use.

Reply:
Manual reference counting is never easy to use. Manual memory management in general is always hard to use. That's why all modern languages have some form of automatic memory management....

At least with this retain/release system it's clear when you should call retain/release. You retain, when you want a new reference for yourself that you hold. You release when you're done with that reference.

With the current inc/dec/del system it's just chaos. You inc when things broke and you dec as soon as possible. You call del when ever, because calling it too much doesn't matter. This sort of chaos is just cause trouble. It's also much easier to create memory leaks. If retain/release a segfault is much more likely than a memory leak and segfaults lead to crashes and issue reports. Memory leaks are silent and easily go undetected.

Anyway, the reference counting is the part of this proposal I'm least fixated on. I'm open to using any solution that works. I just used retain/release, because **I** like it best and I wanted to show that it is a good fit for Elektra (which has been questioned in the past).

Reply:
> AFAIK we have less problems since keyDel is a NOP on wrong use. 

No it is not. "NOP on **wrong** use" is probably impossible and might be equivalent to solving the halting problem.

What you mean is that: Calling `keyDel` on a key with non-zero reference counter is a NOP.
But IMO that is not actually "wrong use" according to the current API and semantics.

In any case, wrong use definitely also includes a case like this:

```c
Key * k = keyNew (...);
keyDel (k);
keyDel (k);
```

And here obviously neither `keyDel` is a NOP.

> free(NULL) is also a NOP (maybe for that reason).

Please stop using this example, it is not a valid comparison. The equivalent for `keyDel` is `keyDel (NULL)`, which works exactly the same as `free (NULL)` (i.e. a NOP). The same would be true for `elektraKeyRelease (NULL)` which would also be a NOP.

> please let clarify such things only in decisions.

I tried in #4676, which went nowhere. Since you didn't even look at the solutions.

And like I said, the ref counting is the one part of the API proposal where I'm most open to other solutions. I just responded to once again, correct your misunderstandings/incorrect statements about the current `keyDel`.

Reply:
> What you mean is that

Important is what you mean, so there should be a very clear description of the problem in a decision.

> Please stop using this example, it is not a valid comparison.

If we are not open to different points of views, we will obviously struggle. I try to understand your viewpoint but failed up to now.

> Since you didn't even look at the solutions.

Because the described problem there (that we need ref counting) is already solved by the current API (=status quo). Decisions must be a problem of status quo, which might exist but is not described. Otherwise we only endlessly refactor based on different tastes.

Reply:
> If we are not open to different points of views, we will obviously struggle.

I **am** open to different points of view. I'm just not open to ignoring or even accepting factually incorrect statements.

I have no pointed out to you on multiple occasions that your comparison of `keyDel` and `free` is flawed. You can compare `free (NULL)` and `keyDel (NULL)`. But you cannot compare `free (NULL)` with `keyDel (k)` where `k` is a key that's already deleted.

> > What you mean is that
> 
> Important is what you mean,

_No_, this was about what (I believe) **you** meant with "keyDel is a NOP on wrong use". You didn't mean "wrong use" in general, you just meant "calling keyDel on a key with non-zero reference counter". At least I assume that is what you mean, since it is the only NOP case for `keyDel` other than the trivial `keyDel (NULL)`.


> Decisions must be a problem of status quo, which might exist but is not described. Otherwise we only endlessly refactor based on different tastes.

Just as a sidenote, no idea to response: Once again, I started decisions for things that are not technically a problem with the status quo, because we are undeniably in the middle of a big API redesign. I already committed to doing that. I don't see why we should not at least think about doing more than strictly necessary, if it might result in further improvements. There is a "rejected" state for decisions. I don't understand, why you refuse to even start a discussion, when there is a clear option of "okay we explored our options, let's keep what we already have".